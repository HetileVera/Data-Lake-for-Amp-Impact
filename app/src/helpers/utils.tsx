import isEmpty from "lodash/isEmpty";
import isNil from "lodash/isNil";
import isString from "lodash/isString";
import numeral from "numeral";

export const delay = (seconds: number) => {
  return new Promise((resolve) => {
    setTimeout(resolve, seconds * 1000);
  });
};

export function humanDay(num: number): string {
  // Account for x11 to x13
  const remainder100 = num % 100;
  if (remainder100 >= 11 && remainder100 <= 13) return `${num}th`;

  const remainder = num % 10;

  if (remainder === 1) return `${num}st`;
  if (remainder === 2) return `${num}nd`;
  if (remainder === 3) return `${num}rd`;

  return `${num}th`;
}

export function humanProcessingTime(seconds: number): string {
  return numeral(seconds).format("00:00:00");
}

export function niceNumber(value: number | string | null | undefined) {
  if (isNil(value)) return "N/A";
  if (isString(value) && isEmpty(value)) return "N/A";
  return numeral(value).format("0,0");
}

export function niceByte(value: number | string | null | undefined) {
  if (isNil(value)) return "N/A";
  if (isString(value) && isEmpty(value)) return "N/A";
  return numeral(value).format("0b");
}

class LocalPersistance {
  // https://developer.mozilla.org/en-US/docs/Web/API/Storage

  supported: boolean = typeof Storage !== "undefined";

  private wrap(fn: () => unknown): unknown {
    if (!this.supported) return;
    try {
      return fn();
    } catch (error) {
      console.error(error);
    }
  }

  clear() {
    this.wrap(() => localStorage.clear());
  }

  getItem(key: string): string | null | undefined {
    return this.wrap(() => localStorage.getItem(key)) as string | null | undefined;
  }

  setItem(key: string, value: string) {
    return this.wrap(() => localStorage.setItem(key, value));
  }

  removeItem(key: string) {
    return this.wrap(() => localStorage.removeItem(key));
  }
}

export const storage = new LocalPersistance();

export function removeSpaces(text: string): string {
  if (isEmpty(text)) return "";

  return text.replace(/^\s+|\s+$/gm,'');
}