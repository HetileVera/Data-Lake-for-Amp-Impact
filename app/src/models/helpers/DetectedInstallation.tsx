import { types, Instance } from "mobx-state-tree";

/**
 * Represents a short information about a detected installation
 */
export const DetectedInstallation = types
  .model("DetectedInstallation", {
    id: "",
    webUrlOrigin: "",
    reachable: true,
    region: "",
    installationJson: types.maybe(types.frozen()),
  })

  .views((self) => ({
    get startDate(): Date | undefined {
      return self.installationJson?.startDate;
    },

    get startedBy(): string | undefined {
      return self.installationJson?.startedBy;
    },

    get accountId(): string | undefined {
      return self.installationJson?.accountId;
    },

    get appFlowConnectionName(): string | undefined {
      return self.installationJson?.appFlowConnectionName;
    },
  }));

export interface IDetectedInstallation extends Instance<typeof DetectedInstallation> {}
