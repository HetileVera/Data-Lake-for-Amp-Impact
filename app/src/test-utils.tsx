import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
