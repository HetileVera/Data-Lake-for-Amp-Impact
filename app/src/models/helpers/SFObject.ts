import filter from "lodash/filter";
import has from "lodash/has";
import sortBy from "lodash/sortBy";

import { types, Instance } from "mobx-state-tree";
import { values } from "mobx";

import { IRawSalesforceObject } from "api/list-salesforce-objects";
import { defaultObjectsMap } from "data/default-npsp-objects";
import { Field, IField } from "models/helpers/Field";
import { isCompoundType } from "models/helpers/utils";

/**
 * Represents a Salesforce object
 */

const sfProps = {
  name: "",
  label: "",
  fields: types.map(Field),
  fieldsLoaded: false,
  selected: false,
  stale: false,
};

const sfActions = (self: any) => ({
  setObject(rawObject: IRawSalesforceObject) {
    self.fields.clear();
    self.name = rawObject.name;
    self.label = rawObject.label;
    self.fieldsLoaded = false;
    self.selected = false;
    self.stale = false;
  },

  setFields(rawFields: { name: string; label: string; type: string }[]) {
    self.fields.clear();
    for (const rawField of rawFields) {
      // Compound fields are excluded
      self.fields.set(rawField.name, { ...rawField, excluded: isCompoundType(rawField.type) });
    }
  },

  setField(rawField: { name: string; label: string; type: string; excluded: boolean }) {
    self.fields.set(rawField.name, { ...rawField, excluded: rawField.excluded || isCompoundType(rawField.type) });
  },

  setFieldsLoaded(flag: boolean) {
    self.fieldsLoaded = flag;
  },

  setSelected(flag: boolean) {
    self.selected = flag;
  },

  markStale() {
    self.stale = true;
  },
});

const sfViews = (self: any) => ({
  get selectableFields(): IField[] {
    // Returns the list of fields that can be selected
    const filtered = filter(values(self.fields), ["canExclude", true]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get listableFields(): IField[] {
    // listable fields are all the fields with the exception of the compound field
    const filtered = filter(values(self.fields), ["compound", false]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get immutableFields(): IField[] {
    const filtered = filter(values(self.fields), { compound: false, canExclude: false });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get selectedFields(): IField[] {
    // Returns all fields that are included (including immutable fields)
    const filtered = filter(values(self.fields), ["excluded", false]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get userSelectedFields(): IField[] {
    // Returns all fields that are included (excluded immutable fields)
    const filtered = filter(values(self.fields), { excluded: false, canExclude: true });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get excludedFields(): IField[] {
    // Returns excluded fields (with the exception of compound fields)
    const filtered = filter(values(self.fields), { excluded: true, compound: false });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get allExcludedFields(): IField[] {
    // Returns all excluded fields including the compound field
    const filtered = filter(values(self.fields), ["excluded", true]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },
  get fieldsCount(): number {
    // All fields (selected or not) excluding compound fields
    return self.listableFields.length;
  },

  get selectedFieldsCount(): number {
    const result = self.selectedFields;
    return result.length;
  },

  get excludedFieldsCount(): number {
    const result = self.excludedFields;
    return result.length;
  },

  get excluded() {
    // Returns the excluded fields as expected by the step functions lambdas. The following is an example of the
    // expected exclude:
    // "exclude": {
    //   "CompoundField__c": { "type": "location", "label": "<field label>", "$comment": "Compound Field" },
    //   "ABC__c": { "type": "string", "label", "<field label>", "$comment": "ABC field" },
    //   "...": { ... },
    // },
    const excludedFields = self.allExcludedFields;
    const exclude: { [index: string]: { type: string; label: string } } = {};
    for (const field of excludedFields) {
      exclude[field.name] = {
        type: field.type,
        label: field.label,
      };
    }

    return exclude;
  },

  get isDefault(): boolean {
    // Is this object one of the default NPSP objects?
    return has(defaultObjectsMap, self.name);
  },
});

export const SFObject = types.model("SFObject", sfProps).actions(sfActions).views(sfViews);

export interface ISFObject extends Instance<typeof SFObject> {}
