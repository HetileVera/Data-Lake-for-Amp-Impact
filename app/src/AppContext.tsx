import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
