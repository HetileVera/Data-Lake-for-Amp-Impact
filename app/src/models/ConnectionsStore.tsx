import { useEffect } from "react";
import isEmpty from "lodash/isEmpty";
import { types, Instance, getParent } from "mobx-state-tree";

import { useInstallation } from "AppContext";
import { listAppflowConnections } from "api/list-appflow-connections";

import { BaseStore, isStoreNew, isStoreError, isStoreReady, isStoreLoading, isStoreReLoading } from "./BaseStore";
import { ICredentials } from "./helpers/Credentials";
import { IInstallation } from "./Installation";

/**
 * Represents the store for the AppFlow Salesforce connection names
 */
export const ConnectionsStore = BaseStore.named("ConnectionsStore")
  .props({
    connectionNames: types.array(types.string),
  })

  .views((self) => ({
    get empty() {
      return isEmpty(self.connectionNames);
    },

    get credentials(): ICredentials {
      const parent: IInstallation = getParent(self);
      return parent.credentials;
    },

    get region(): string {
      const parent: IInstallation = getParent(self);
      return parent.region;
    },
  }))

  .actions((self) => {
    const superReset = self.reset;
    return {
      reset() {
        superReset();
        self.connectionNames.clear();
      },

      async doLoad() {
        const accessKey = self.credentials.accessKey;
        const secretKey = self.credentials.secretKey;
        const region = self.region;

        const rawConnectionNames = await listAppflowConnections({ accessKey, secretKey, region });
        self.runInAction(() => {
          self.connectionNames.replace(rawConnectionNames);
        });
      },
    };
  });

// see https://mobx-state-tree.js.org/tips/typescript
export interface IConnectionsStore extends Instance<typeof ConnectionsStore> {}

export function useConnectionsStore() {
  const installation = useInstallation();
  const store = installation.connectionsStore;

  useEffect(() => {
    if (!isStoreNew(store)) return;
    store.load();
  }, [store]);

  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
