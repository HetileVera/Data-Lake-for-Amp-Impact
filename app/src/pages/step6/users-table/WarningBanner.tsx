import React from "react";
import { observer } from "mobx-react";
import { Alert, Box, AlertDescription, Button, AlertTitle, useToast, CloseButton } from "@chakra-ui/react";

import { useInstallation } from "AppContext";
import { useColorScheme } from "models/useColorScheme";
import { delay } from "helpers/utils";
import { IUser, TUserAccessStatus } from "models/helpers/User";
import { useUsersStore } from "models/useUsersStore";

interface IWarningBanner {
  user: IUser;
}

export const WarningBanner = observer(({ user }: IWarningBanner) => {
  const { colorScheme } = useColorScheme();
  const installation = useInstallation();
  const { store } = useUsersStore(installation);
  const [processing, setProcessing] = React.useState(false);
  const [error, setError] = React.useState("");
  const toast = useToast();

  const handleAttachPolicy = async () => {
    setError("");
    setProcessing(true);

    try {
      await delay(1); // To give a better visual cue
      await store.attachAthenaPolicy(user, installation);
      toast({
        status: "success",
        description: "Policy attached",
        isClosable: false,
        duration: 1000,
      });
    } catch (err: any) {
      console.log(err);
      setError(err.message);
    }
    setProcessing(false);
  };

  let content = null;
  if (user.hasAccess) return null;
  if (user.accessStatus === TUserAccessStatus.No_Keys) {
    // @ts-ignore
    content = (
      <Alert status="warning" mt={0} mb={4} borderRadius="md" color="red.700" alignItems="flex-start" fontSize="sm">
        <AlertDescription>
          The user does not have access to the data lake because there are no access keys associated with the user. Create an access key for the user
          using the button below.
        </AlertDescription>
      </Alert>
    );
  }

  if (user.accessStatus === TUserAccessStatus.No_Active_Keys) {
    // @ts-ignore
    content = (
      <Alert status="warning" mt={0} mb={4} borderRadius="md" color="red.700" alignItems="flex-start" fontSize="sm">
        <AlertDescription>
          The user does not have access to the data lake because all the access keys are inactive. Activate one for the user to gain access.
        </AlertDescription>
      </Alert>
    );
  }

  if (user.accessStatus === TUserAccessStatus.No_POLICY) {
    // @ts-ignore
    content = (
      <Alert status="warning" mt={0} mb={4} borderRadius="md" color="red.700" alignItems="flex-start" fontSize="sm">
        <Box>
          <AlertDescription>
            The user does not have access to the data lake because the necessary policy is not attached to the user. Click the 'Attach Policy' button
            to attempt to remedy the problem.
          </AlertDescription>
          <Box textAlign="right" mt={4}>
            <Button colorScheme={colorScheme} size="xs" isLoading={processing} loadingText="Attaching Policy" onClick={handleAttachPolicy}>
              Attach Policy
            </Button>
          </Box>
        </Box>
      </Alert>
    );
  }

  return (
    <>
      {error && (
        <Alert status="error" variant="left-accent" mt={0} mb={4} color="red.700" alignItems="flex-start" fontSize="sm">
          <Box flex="1">
            <AlertTitle>Error</AlertTitle>
            <AlertDescription display="block">{error}</AlertDescription>
          </Box>
          <CloseButton position="absolute" right="8px" top="8px" onClick={() => setError("")} />
        </Alert>
      )}

      {content}
    </>
  );
});
