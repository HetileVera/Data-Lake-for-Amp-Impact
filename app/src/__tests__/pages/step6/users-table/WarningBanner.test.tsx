import { cleanup, fireEvent, screen, waitFor } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, CustomChakraProvider } from "test-utils";

import * as appContext from "AppContext";
import * as usersStore from "models/useUsersStore";

import { WarningBanner } from "pages/step6/users-table/WarningBanner";
import { IUser, User } from "models/helpers/User";
import { IRawAccessKey, TAccessKeyStatus } from "models/helpers/UserAccessKey";

describe("WarningBanner component", () => {
  beforeEach(() => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        id: "1",
        startDate: "2022-01-01",
        startedBy: "test-user",
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        credentials: [],
        nextStepNumber: "2",
        createUsersStep: {
          markStarted: jest.fn(),
          markCompleted: jest.fn(),
        },
      };
    });

    // @ts-ignore
    jest.spyOn(usersStore, "useUsersStore").mockImplementation(() => {
      return {
        store: {
          attachAthenaPolicy: (user, installation) => {
            user.hasAthenaManagedPolicy = true;
            return true;
          },
        },
      };
    });
  });

  afterEach(() => {
    cleanup;
  });

  test("snapshot", () => {
    const user: IUser = User.create({
      id: "1",
      name: "Simsek Mert",
      arn: "arn:aws:iam::xyzxyzxyzxx:root",
      createDate: new Date("2022-01-05"),
      accessKeys: {},
      policies: ["some-test-policy"],
      hasAthenaManagedPolicy: false,
      stale: false,
    });

    const tree = create(
      <CustomChakraProvider>
        <WarningBanner user={user} />
      </CustomChakraProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render with a user having access", () => {
    const user: IUser = User.create({
      id: "1",
      name: "Simsek Mert",
      arn: "arn:aws:iam::xyzxyzxyzxx:root",
      createDate: new Date("2022-01-05"),
      accessKeys: {},
      policies: ["some-test-policy"],
      hasAthenaManagedPolicy: true, // this makes sure the user has access
      stale: false,
    });

    render(<WarningBanner user={user} />);

    expect(screen.getByTestId("test-root")).toBeTruthy();
  });

  test("render with a user with no policy", () => {
    const user: IUser = User.create({
      id: "1",
      name: "Simsek Mert",
      arn: "arn:aws:iam::xyzxyzxyzxx:root",
      createDate: new Date("2022-01-05"),
      accessKeys: {},
      policies: ["some-test-policy"],
      hasAthenaManagedPolicy: false,
      stale: false,
    });

    render(<WarningBanner user={user} />);

    expect(screen.getByText(/necessary policy is not attached/i)).toBeTruthy();
  });

  test("render with a user with no keys", () => {
    const user: IUser = User.create({
      id: "1",
      name: "Simsek Mert",
      arn: "arn:aws:iam::xyzxyzxyzxx:root",
      createDate: new Date("2022-01-05"),
      accessKeys: {},
      policies: ["arn:aws:iam::aws:policy/AdministratorAccess"],
      hasAthenaManagedPolicy: false,
      stale: false,
    });

    render(<WarningBanner user={user} />);

    expect(screen.getByText(/no access keys associated with the user/i)).toBeTruthy();
  });

  test("render with a user with inactive keys", () => {
    const user: IUser = User.create({
      id: "1",
      name: "Simsek Mert",
      arn: "arn:aws:iam::xyzxyzxyzxx:root",
      createDate: new Date("2022-01-05"),
      accessKeys: {},
      policies: ["arn:aws:iam::aws:policy/AdministratorAccess"],
      hasAthenaManagedPolicy: false,
      stale: false,
    });
    const rawAccessKey: IRawAccessKey = {
      id: "1",
      secret: "some-test-secret",
      createDate: new Date("2022-02-01"),
      status: TAccessKeyStatus.Inactive,
    };

    // Set the raw access key
    user.setAccessKey(rawAccessKey);

    render(<WarningBanner user={user} />);

    expect(screen.getByText(/access keys are inactive/i)).toBeTruthy();
  });

  test("attach policy for a user with no policy", async () => {
    const user: IUser = User.create({
      id: "1",
      name: "Simsek Mert",
      arn: "arn:aws:iam::xyzxyzxyzxx:root",
      createDate: new Date("2022-01-05"),
      accessKeys: {},
      policies: ["some-test-policy"],
      hasAthenaManagedPolicy: false,
      stale: false,
    });

    render(<WarningBanner user={user} />);

    expect(screen.getByText(/necessary policy is not attached/i)).toBeInTheDocument();

    const button = screen.getByRole("button", { name: "Attach Policy" });
    fireEvent.click(button);

    await screen.findByText("Attaching Policy");
    await screen.findByText("Attach Policy");
  });
});
