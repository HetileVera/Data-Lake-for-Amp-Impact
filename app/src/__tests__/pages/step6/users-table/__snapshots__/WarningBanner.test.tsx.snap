// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarningBanner component snapshot 1`] = `
Array [
  <div
    data-testid="test-root"
  >
    
    <div
      className="chakra-alert css-1l0fb9r"
      role="alert"
    >
      <div
        className="css-0"
      >
        <div
          className="chakra-alert__desc css-zzks76"
        >
          The user does not have access to the data lake because the necessary policy is not attached to the user. Click the 'Attach Policy' button to attempt to remedy the problem.
        </div>
        <div
          className="css-oo7t1"
        >
          <button
            className="chakra-button css-1c6k3b0"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Attach Policy
          </button>
        </div>
      </div>
    </div>
  </div>,
  <span />,
]
`;
